feat: implement EditOrganizerScreen

Zeleph25 opened the pull request:
**Description**  
This PR adds the `EditOrganizerScreen` UI, allowing owners to edit their organization's details. It reuses existing form components (`FormTextField`, `PrefixPhoneRow`, `SubmitButton`) and uses hardcoded data for now.

⚠️ **Important Considerations:**  
This PR includes **unit and UI tests** and many **imports**, so the current lines of code are higher than expected. The high LOC is therefore not an indication of unnecessary code, but a result of proper testing and dependencies.

**Purpose**  
Provide a simple, consistent interface for updating organization information.

**Issue**  
Edit Organization Screen #86

**Trade-offs**
- Image/logo upload support is not included in this PR.
- Some form logic and components are duplicated between the **Become Organizer** and **Edit Organizer** screens; a future refactor should extract shared form code into a reusable utility.

**Acceptance Criteria**
- [ ] All organization fields are editable.
- [ ] Changes are saved via the Submit button.
- [ ] No visual or functional regressions compared to the Become Organizer screen.
- [ ] Errors are displayed appropriately via snackbar messages.

**How to test**
- Run unit/UI tests: **EditOrganizationScreenTest** and **EditOrganizationViewModelTest**  
- Open `OrganizerNavigationPreview()` to manually check the UI  
- Verify all fields are visible, editable, and validated correctly  
- Test dropdown behavior (country prefix) and scrolling on smaller screens  
- Confirm the Submit button triggers the update action  
- Check visual consistency with the Become Organizer screen in both light & dark mode  

**Since the preview code to test the screen is lengthy, it is collapsed below. Click Expand to view it, then copy-paste it into your preview file to test the screen:**

### Preview Image
<table>
  <tr>
    <td align="center">
<img width="286" height="598" alt="Edit Organization" src="https://github.com/user-attachments/assets/b67664d8-156c-4da9-abea-f2016e88500c" />      <br>
      <em> Edit Organization </em>
    </td>
  </tr>
</table>

<details>
<summary><strong>Preview Code (click to expand)</strong></summary>

```kotlin
@OptIn(ExperimentalMaterial3Api::class)
@Preview(showBackground = true)
@Composable
fun OrganizerNavigationPreview() {
    OnePassTheme {
        val currentUserId = "previewUser"
        val createViewModel = remember { BecomeOrganizerViewModel() }
        val editViewModel = remember { EditOrganizationViewModel() }

        var screenState by remember { mutableStateOf<ScreenState>(ScreenState.Menu) }
        var currentOrganizationId by remember { mutableStateOf<String?>(null) }
        var organizations by remember { mutableStateOf(listOf<Organization>()) }
        var viewingOrgList by remember { mutableStateOf(false) }

        Surface(modifier = Modifier.fillMaxSize(), color = MaterialTheme.colorScheme.background) {
            when {
                screenState == ScreenState.Menu && !viewingOrgList -> {
                    Column(
                        modifier = Modifier.fillMaxSize().padding(16.dp),
                        verticalArrangement = Arrangement.Center) {
                        Text(
                            "Organizer Actions",
                            style = MaterialTheme.typography.headlineMedium,
                            color = MaterialTheme.colorScheme.onBackground)
                        Spacer(Modifier.height(32.dp))

                        Button(
                            onClick = { screenState = ScreenState.Create },
                            modifier = Modifier.fillMaxWidth(),
                            colors =
                                ButtonDefaults.buttonColors(
                                    containerColor = MaterialTheme.colorScheme.primary)) {
                            Text("Create Organizer", color = MaterialTheme.colorScheme.onPrimary)
                        }

                        Spacer(Modifier.height(16.dp))

                        if (organizations.isNotEmpty()) {
                            Button(
                                onClick = { viewingOrgList = true },
                                modifier = Modifier.fillMaxWidth(),
                                colors =
                                    ButtonDefaults.buttonColors(
                                        containerColor = MaterialTheme.colorScheme.primary)) {
                                Text(
                                    "View Existing Organizations",
                                    color = MaterialTheme.colorScheme.onPrimary)
                            }
                        }
                    }
                }
                screenState == ScreenState.Menu && viewingOrgList -> {
                    Column(
                        modifier =
                            Modifier.fillMaxSize().padding(16.dp).verticalScroll(rememberScrollState())) {
                        Text(
                            "Your Organizations:",
                            style = MaterialTheme.typography.headlineMedium,
                            color = MaterialTheme.colorScheme.onBackground)
                        Spacer(Modifier.height(16.dp))

                        organizations.forEach { org ->
                            Button(
                                onClick = {
                                    currentOrganizationId = org.id
                                    screenState = ScreenState.Edit
                                    viewingOrgList = false
                                },
                                modifier = Modifier.fillMaxWidth().padding(vertical = 4.dp),
                                colors =
                                    ButtonDefaults.buttonColors(
                                        containerColor = MaterialTheme.colorScheme.secondaryContainer)) {
                                Text(org.name, color = MaterialTheme.colorScheme.onSecondaryContainer)
                            }
                        }

                        Spacer(Modifier.height(16.dp))
                        Button(
                            onClick = { viewingOrgList = false },
                            modifier = Modifier.fillMaxWidth(),
                            colors =
                                ButtonDefaults.buttonColors(
                                    containerColor = MaterialTheme.colorScheme.secondaryContainer)) {
                            Text("Back", color = MaterialTheme.colorScheme.onSecondaryContainer)
                        }
                    }
                }
                screenState == ScreenState.Create -> {
                    Scaffold(
                        topBar = {
                            TopAppBar(
                                title = { Text("Create Organizer") },
                                navigationIcon = {
                                    IconButton(onClick = { screenState = ScreenState.Menu }) {
                                        Icon(
                                            androidx.compose.material.icons.Icons.Default.ArrowBack,
                                            contentDescription = "Back",
                                            tint = MaterialTheme.colorScheme.onPrimary)
                                    }
                                },
                                colors =
                                    TopAppBarDefaults.topAppBarColors(
                                        containerColor = MaterialTheme.colorScheme.primary,
                                        titleContentColor = MaterialTheme.colorScheme.onPrimary))
                        }) { padding ->
                        Box(modifier = Modifier.padding(padding)) {
                            BecomeOrganizerScreen(
                                ownerId = currentUserId,
                                viewModel = createViewModel,
                                onOrganizationCreated = { newOrgId ->
                                    val name = createViewModel.formState.value.name.value
                                    organizations =
                                        organizations +
                                                Organization(id = newOrgId, name = name, ownerId = currentUserId)
                                    currentOrganizationId = newOrgId
                                    screenState = ScreenState.Menu
                                })
                        }
                    }
                }
                screenState == ScreenState.Edit -> {
                    currentOrganizationId?.let { orgId ->
                        Scaffold(
                            topBar = {
                                TopAppBar(
                                    title = { Text("Edit Organizer") },
                                    navigationIcon = {
                                        IconButton(onClick = { screenState = ScreenState.Menu }) {
                                            Icon(
                                                androidx.compose.material.icons.Icons.Default.ArrowBack,
                                                contentDescription = "Back",
                                                tint = MaterialTheme.colorScheme.onPrimary)
                                        }
                                    },
                                    colors =
                                        TopAppBarDefaults.topAppBarColors(
                                            containerColor = MaterialTheme.colorScheme.primary,
                                            titleContentColor = MaterialTheme.colorScheme.onPrimary))
                            }) { padding ->
                            Box(modifier = Modifier.padding(padding)) {
                                EditOrganizationScreen(organizationId = orgId, viewModel = editViewModel)
                            }
                        }
                    }
                }
            }
        }
    }
}

sealed class ScreenState {
    object Menu : ScreenState()

    object Create : ScreenState()

    object Edit : ScreenState()
}
</details>

BruceKZ reviewed (changes requested):
Nice work Adam @Zeleph25!

I reviewed your PR structure, and here are a few points you might want to consider:

1. Rebase the branch `85-become-organizer-view-model` on the latest `main` to fix the permission issue in CI.
2. Rebase the branch `86-edit-organizer-screen` on `85-become-organizer-view-model`, and change the PR base to `85`. After that, your VM will no longer appear as new code in the “Files changed” section.  
   P.S. Once `85` is merged into `main`, you can rebase `86` on `main` again and update the PR base to `main`. That way, the PR will correctly show only the new code.
3. It’s generally recommended to include thorough tests for any new code in each PR to ensure everything merged into `main` is properly verified. However, since we need to merge the screen early to support navigation work, please make sure to add the missing tests later.

I also have some code-level comments, but I’ll leave them as inline code comments after you complete the rebase, so they can be addressed in the correct PR.

Zeleph25 commented:
This PR is now open again for review.  

Thanks @BruceKZ  for the comment you left above! I’ve done the rebases as suggested:

- `85-become-organizer-view-model` is rebased on the latest `main` to fix the CI permission issue.  
- `86-edit-organizer-screen` is now rebased on `main`, and the PR base is set to `main` since `85-become-organizer-view-model` has already been merged. This ensures the VM changes no longer show up as new code in the “Files changed” section.

I’ve also added the missing tests for the new screen, as suggested.  

Looking forward to your inline comments. Thanks again for the guidance!

BruceKZ commented on app/src/main/java/ch/onepass/onepass/ui/organizer/EditOrganizerScreen.kt:
```diff
@@ -0,0 +1,176 @@
+    // Load the first organization for the user
+    // when the screen is first composed
+    viewModel.loadFirstOrganizationForUser(userId)
+  }
```
It is better to compose the organization with the given `organizationId` instead of the first organization for the user since this screen is for editing an organization.

BruceKZ commented on app/src/main/java/ch/onepass/onepass/ui/organizer/EditOrganizerScreen.kt:
```diff
@@ -0,0 +1,176 @@
+                              facebook = formState.facebook.value.ifBlank { null },
+                              tiktok = formState.tiktok.value.ifBlank { null },
+                              address = formState.address.value.ifBlank { null })
+                      viewModel.updateOrganization(data)
```
The method `updateOrganization` might throw error but the UI didn't handle the error. My suggestion is to use `try-catch` here.

Zeleph25 replied:
Thanks @BruceKZ, I added a try-catch around `updateOrganization` to properly handle any errors in the UI. You can see the changes either in my other comment about `updateOrganization` or directly in the updated code.

BruceKZ commented on app/src/main/java/ch/onepass/onepass/ui/organizer/EditOrganizerScreen.kt:
```diff
@@ -0,0 +1,176 @@
+      coroutineScope.launch { snackbarHostState.showSnackbar(it) }
+      formViewModel.clearError()
+    }
+  }
```
The `LaunchedEffect(uiState)` listens to `formViewModel.uiState` (from `BecomeOrganizerViewModel`), which is designed for the "create organization" flow. It checks for `successOrganizationId` which doesn't make sense in an edit context. The edit screen should have its own success/error handling or use the `EditOrganizationViewModel`'s state.

BruceKZ reviewed (changes requested):
Great work!

Overall, this PR follows the MVVM architecture well — the separation between the UI and ViewModel layers is clear, and the implementation of the “edit organization” feature is solid.  
That said, there are still a few issues to address:

1. **Naming**  
   In our product design, there is no item called “Organizer.” The correct term should be **“Organization.”**  
   To clarify: users can create or join organizations, and every event’s creator (or initiator) must be an organization — not an individual user (even if that organization only has one member).  
   It’s good that you separated related logic using `organization` and `organizer`, but since “Organizer” doesn’t exist in our domain model, please replace it throughout the UI and code where applicable.

2. **Edit Screen Logic**  
   The edit screen logic is quite similar to what we built in the Bootcamp sessions for addTodo/editTodo.  
   As an edit screen, it should pre-fill the form with existing values based on the `organizationId`.  
   The function `loadFirstOrganizationForUser` doesn’t fit well here — if the screen receives an `organizationId`, it should clearly indicate that it’s editing the organization corresponding to that ID.

3. **Loading State**  
   Building on the previous point, if you decide to preload data using the `organizationId`, please include a loading indicator (spinner or skeleton view) to show the user that data is being fetched.

There are also a few more issues noted in the inline comments.

Zeleph25 commented:
Thanks @BruceKZ  for the review. I’ve gone through all the remarks and addressed them. There was one suggestion for future PRs, but all the required fixes—from minor to more critical issues—have been implemented. I also added appropriate tests to cover the changes. Is everything good now?

BruceKZ reviewed (changes requested):
GREAT AND QUICK FIX! THANK YOU ADAM @Zeleph25 !

After reviewing your latest code, I found some small issues which could be fixed quickly. There are 3 comments in the code, please check them.

Zeleph25 commented on app/src/main/java/ch/onepass/onepass/ui/organizer/EditOrganizerScreen.kt:
```diff
@@ -0,0 +1,223 @@
+      org.contactPhone?.let {
+        val prefix = it.takeWhile { c -> c == '+' || c.isDigit() }
+        prefixDisplayText = prefix
+      }
```
Thanks for the comment! In this case, the current logic is intentional and aligns with how the value is stored.

We store `contactPhone` in Firestore already in the full E.164 format (e.g. `+41998887766`). Since the number is stored with the prefix included, extracting only the region prefix is not required at this stage. The UI should simply show the saved value back to the user as-is.

So this code is correct for our current data format:

```kotlin
org.contactPhone?.let {
    val prefix = it.takeWhile { c -> c == '+' || c.isDigit() }
    prefixDisplayText = prefix
}
```

If in the future we decide to store the country prefix and the phone number separately, then the logic should be updated. For now, storing the full E.164 number in the database is the intended behaviour and matches how users enter the value.

Zeleph25 commented on app/src/main/java/ch/onepass/onepass/ui/organizer/EditOrganizationViewModel.kt:
```diff
@@ -0,0 +1,156 @@
+                tiktok = data.tiktok,
+                address = data.address,
+                createdAt = currentOrg.createdAt ?: Timestamp.now() // fallback if null
+                )
```
Yes, you're right — using `copy()` is better. I’ve updated the code to use it.

BruceKZ reviewed (approved):
Great job, @Zeleph25 ! Thanks for the quick turnaround.
Everything looks good to me — I’m approving the PR.
